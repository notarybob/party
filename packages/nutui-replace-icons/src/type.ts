export interface IOptions {
  sourceLibrary: string[]
  targetLibrary: string
  iconMappings?: { [key: string]: string }
}
