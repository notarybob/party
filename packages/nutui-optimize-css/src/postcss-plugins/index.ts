import postcss, { ProcessOptions, Root, Document } from 'postcss'
import { merge } from 'lodash'
import cssVariables from 'postcss-css-variables'
import { parse } from 'postcss-scss'
import * as fs from 'fs'

export type CSSVariable = `--${string}`
export type FilePath = string

export interface Options {
  removeRtl: boolean
  cssVariables: {
    include: FilePath[]
    exclude: CSSVariable[]
    type: 'normal' | 'replace'
  }
}

function removeRtl(rule: any, canRemove: boolean) {
  if (!canRemove) return

  const sourceFile = rule.source.input.file
  if (
    sourceFile &&
    sourceFile.indexOf('@nutui') === -1 &&
    sourceFile.indexOf('@dongdesign') === -1
  )
    return
  if (
    rule.selector.indexOf('nut-rtl') > -1 ||
    rule.selector.indexOf('[dir=rtl]') > -1
  )
    rule.remove()
}

async function replaceCssVariables(
  root,
  cssVariablesContent: string[],
  exclude: string[] = []
) {
  cssVariablesContent.push(root.toResult().css)
  const options: ProcessOptions<Document | Root> = {
    parser: parse,
    from: undefined,
  } as ProcessOptions<Root>
  const replacedCss = postcss([
    cssVariables({
      preserve: (declaration) => {
        if (exclude.includes(declaration.prop)) {
          return true
        }
        const cssvars = declaration.value.match(/var\((--nutui-[\w\d-]+)\)/)
        if (cssvars && exclude.includes(cssvars[1])) return true
        return false
      },
    }),
  ]).process(cssVariablesContent.join('\n'), options).css

  const replacedRoot = postcss.parse(replacedCss)
  root.raws = replacedRoot.raws
  root.nodes = replacedRoot.nodes
}

export function optimizeCss(opts: Options) {
  const defaultConfig = {
    removeRtl: false,
    cssVariables: {
      include: [],
      type: 'normal',
    },
  }
  const config = merge(defaultConfig, opts)
  const cssVariablesContent: string[] = []
  if (config.cssVariables.type !== 'normal') {
    config.cssVariables.include.forEach((p: string) => {
      let content = ''
      try {
        // 从绝对路径读取 CSS 变量的内容
        content = fs.readFileSync(p).toString()
      } catch (e) {
        content = ''
      }
      cssVariablesContent.push(content)
    })
  }

  return {
    postcssPlugin: 'postcss-optimize-css',
    OnceExit(root) {
      if (config.cssVariables.type === 'replace') {
        replaceCssVariables(
          root,
          cssVariablesContent,
          config.cssVariables.exclude
        )
      }
    },
    Rule(rule) {
      removeRtl(rule, opts.removeRtl)
    },
  }
}
