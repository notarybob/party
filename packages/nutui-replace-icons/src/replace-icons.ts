import { IOptions } from './type'

function replace(options: IOptions) {
  const sourceLibrary = options.sourceLibrary || []
  const targetLibrary = options.targetLibrary
  return {
    visitor: {
      ImportDeclaration(path) {
        // import {Loading} from '@nutui/icons-react'
        if (sourceLibrary.indexOf(path.node.source.value) > -1) {
          // 替换包名称
          path.node.source.value = targetLibrary
          path.node.specifiers.forEach((specifier) => {
            // 根据 iconMappings 进行替换
            const iconMappings = options.iconMappings
            if (iconMappings && iconMappings[specifier.imported.name]) {
              specifier.imported.name = iconMappings[specifier.imported.name]
            }
          })
        } else {
          sourceLibrary.forEach((library) => {
            const libraryPattern = new RegExp(`^${library}(?:/|$)`)
            if (libraryPattern.test(path.node.source.value)) {
              // import '@nutui/icons-react-taro/dist/style_iconfont.css'
              path.node.source.value = path.node.source.value.replace(
                library,
                targetLibrary
              )
            }
          })
        }
      },
    },
  }
}

export function replaceIcons(
  options: IOptions = {
    sourceLibrary: ['@nutui/icons-react', '@nutui/icons-react-taro'],
    targetLibrary: '',
  }
) {
  if (!options.targetLibrary) {
    return {}
  }
  return replace(options)
}
