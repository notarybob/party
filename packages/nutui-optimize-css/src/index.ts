import { optimizeCss } from './postcss-plugins'

export default optimizeCss
