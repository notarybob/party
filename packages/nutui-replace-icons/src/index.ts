import { replaceIcons } from './replace-icons'

export default replaceIcons
