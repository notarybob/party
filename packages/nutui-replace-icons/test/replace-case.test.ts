import babel from '@babel/core'
import { describe, expect, it } from 'vitest'
import { replaceIcons } from '../src/replace-icons'

const plugin = replaceIcons({
  sourceLibrary: ['@nutui/icons-react', '@nutui/icons-react-taro'],
  targetLibrary: '@test/aa',
  iconMappings: {
    Loading: 'Star',
  },
})

const babelOptions = {
  presets: ['@babel/preset-react'],
  plugins: [plugin],
}
const caseIns = `
import { Loading } from '@nutui/icons-react'
import { ArrowSize6 as Arrow } from '@nutui/icons-react'
const ReplaceOne = () => {
  return <><Loading /> <Arrow /></>
}
`
describe('', () => {
  it('replace Loading icons with Star', () => {
    const ast = babel.transformSync(caseIns, babelOptions)
    // @ts-ignore
    expect(ast.code).toMatchSnapshot()
  })
})
