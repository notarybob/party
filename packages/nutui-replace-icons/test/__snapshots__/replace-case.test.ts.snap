// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`replace Loading icons with Star 1`] = `
"import { Star as Loading } from "@test/aa";
import { ArrowSize6 as Arrow } from "@test/aa";
const ReplaceOne = () => {
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Loading, null), " ", /*#__PURE__*/React.createElement(Arrow, null));
};"
`;
